### PR TITLE
Customizer Settings Snapshot

### PR DESCRIPTION
**Pull request by valendesigns**
Basically, this is Customizer Transactions lite which would allow a user to share a customizer snapshot with anyone.
#### User Stories
- [x] Share the exact current state of the Customizer settings, including dirty and non-dirty settings, a “full snapshot”.
- [x] Share the modified (dirty) settings, for a “diff snapshot”.
- [x] Allow unauthenticated user to access a snapshot to view on the frontend, or authenticated users to load the snapshot in the Customizer (maybe there should be an admin bar link for this instead of automatically redirecting).
#### Todo's
- [x] Build a custom post type that saves a JSON blob of all the Customizer settings into post_content to be played back at a future date.
- [x] Potentially, we could have an option to save only the dirty settings. This means any changes made after the snapshot was created would persist and the dirty settings would be played on top of the current settings.
- [x] Use `customize_snapshot_uuid` as the query parameter, which is passed a UUID mapped to a specific post title.
- [x] Create a new UUID every time Share is clicked, if the user is not viewing a saved snapshot.
- [x] Add a button to the Customizer that indicates the user can share the current URL with non authenticated users to preview.
- [x] Clicking "Save" when previewing a snapshot should save & publish the snapshot and save the Customizer, while clicking "Save" when _not_ previewing a snapshot should only save the Customizer.
- [x] Unit test `Customize_Snapshot_Manager`.
- [x] Unit test `Customize_Snapshot`.
- [x] Add Underscore templates for the dialogs.
- [ ] Preview a snapshot when a UUID is present in the query string.
  - [x] Rewrite the Customize deep link.
  - [x] Don't allow snapshots when previewing an inactive theme or for a different theme than the snapshot was intended for.
  - [x] Update values for authenticated users in the Customizer.
  - [x] Update values for all users by Bootstrapping a headless Customizer on the front-end.
  - [ ] Disable the navigation click events when visiting the preview URL.
  - [ ] Make it work with the [contextual-settings](https://github.com/xwp/wp-customize-contextual-settings) plugin.
#### Known Issues
- [x] **Blocker**: When `WP_Customize_Manager::add_dynamic_settings` is called from `Customize_Snapshot_Manager::save`, via the save or update snapshot methods, errors are thrown if you've added widgets or menus. Based on the errors the issue stems from `WP_Customize_Manager::get_setting` being passed an illegal offset or `null`.
- [x] **Bug**: Site icon doesn't preview.
- [x] **Bug**: Nav Menu doesn't preview.
#### Caveats

Currently the saved data is JS sanitized. There is a risk with that, that if the site changes their auth key constant, that the JS values for widgets would then fail to convert from JS values back to PHP values, because the HMAC check would then fail. However, I have not found a way to save them as sanitized PHP and then convert them back to JS at run time without the preview breaking.

### CODE REVIEW

File: php/class-plugin.php [hunk 1]
```diff
@@ -163,6 +169,9 @@ function init() {
 		if ( $this->is_module_active( 'deferred_customize_widgets' ) ) {
 			$this->deferred_customize_widgets = new Deferred_Customize_Widgets( $this );
 		}
+		if ( $this->is_module_active( 'customize_settings_snapshot' ) ) {
+			$this->customize_settings_snapshot = new Customize_Settings_Snapshot( $this );
```
Comment (westonruter) [inline php/class-plugin.php]: @valendesigns Given how the transaction patch is basically two parts, the extensions to the `WP_Customize_Manager` and then a separate `WP_Customize_Transaction` that is instantiated by the former, maybe it would make sense to take the same approach here, where we can have a `Customize_Snapshot_Manager` (or `Customize_Snapshots` or `Customize_Settings_Snapshotting`?) and a `Customize_Settings_Snapshot`. This would allow our code here in the plugin to more closely mirror the transactions patch, and make it easier to unit-test since we could more easily instantiate `Customize_Settings_Snapshot` in isolation.

Comment (valendesigns) [inline php/class-plugin.php]: The Transactions patch directly changes `WP_Customize_Manager`, which we are not able to do here and cannot be extended as it's a `final` class. I'm all for isolating the functionality though the approach the Transaction patch took is not exactly possible. I guess I'm not completely following how we could make a `Customize_Snapshot_Manager` work like the patch does, but I'm happy to take another look and pivot on the PR. In your mind what would the responsibility of the `Customize_Snapshot_Manager` class be?

Comment (westonruter) [inline php/class-plugin.php]: `Customize_Snapshot_Manager` would be doing the same things for `Customize_Settings_Snapshot` that `WP_Customize_Manager` does for `WP_Customize_Transaction`. In other words, the manager class would instantiate the `Customize_Settings_Snapshot` class, it would pass values into it via Ajax and get values out of it. The `Customize_Settings_Snapshot` would be standalone and more agnostic to its environment, like `WP_Customize_Setting` is.

Comment (valendesigns) [inline php/class-plugin.php]: OK, I'll make some changes to the structure before I move forward with the functionality.

**Comment by valendesigns** [comment]
@westonruter Do the latest round of changes address your concerns?

**Comment by westonruter** [comment]
@valendesigns yes, I think that organization is better. Thanks!

**Comment by valendesigns** [comment]
@westonruter I've got a couple questions.
- What is the behavior of the Share button meant to be when you click it? Obviously it saves the snapshot, but I'm not certain on how it should interact with the user. Does it alert/prompt the user with a preview URL and `customize_snapshot_uuid`? Or is there some other option I've overlooked?
- I would like to add some unit tests for `Customize_Snapshot_Manager` later today when I get up, but I'm not certain how best to bootstrap the class. Could you possibly write a test so I can see how you're setting it up?

**Comment by westonruter** [comment]
@valendesigns:

> What is the behavior of the Share button meant to be when you click it? Obviously it saves the snapshot, but I'm not certain on how it should interact with the user. Does it alert/prompt the user with a preview URL and `customize_snapshot_uuid`? Or is there some other option I've overlooked?

Good question. Yeah, I think clicking the snapshot link should cause the spinner to show while the Ajax request is being made, and then a dialog can appear with the URL to the _frontend_ previewed URL with the UUID query param added.

Oh something else is that the Ajax response should include another UUID to use for the next time that the link is clicked, as I don't think you'd want to overwrite the existing snapshot if you click the share link again? (That is, unless the UUID is generated in JavaScript instead.)

> I would like to add some unit tests for `Customize_Snapshot_Manager` later today when I get up, but I'm not certain how best to bootstrap the class. Could you possibly write a test so I can see how you're setting it up?

Probably best to look at how other modules are bootstrapped for unit testing in the Customize Widgets Plus plugin.

**Comment by valendesigns** [comment]
> Oh something else is that the Ajax response should include another UUID to use for the next time that the link is clicked, as I don't think you'd want to overwrite the existing snapshot if you click the share link again? (That is, unless the UUID is generated in JavaScript instead.)

I was under the impression that a UUID is basically attached to the Customizer session and if you clicked the share link multiple times in a single session it would just update the current snapshot. What you're saying is that you can create multiple snapshots in a single Customizer session?

> Probably best to look at how other modules are bootstrapped for unit testing in the Customize Widgets Plus plugin.

I did check out the other modules, I'll take another crack at it after I finish up with the `dialog`.

**Comment by westonruter** [comment]
@valendesigns:

> I was under the impression that a UUID is basically attached to the Customizer session and if you clicked the share link multiple times in a single session it would just update the current snapshot. What you're saying is that you can create multiple snapshots in a single Customizer session?

I was originally thinking the UUID would be attached to the session as well, as this is how transactions work. But since we're not implementing transactions but a different use case, then maybe it would make sense to be able to create multiple snapshots from a single session. 

I think the other use case, where the session _is_ tied to a transaction, would make more sense with a paradigm where the user is intending to draft settings for publishing. So in that scenario, we could separate out the “Save & Publish” button into two separate “Save” and “Publish” buttons, here hitting “Save” would just update an existing snapshot (in the transactions patch, every change updates the existing transaction post, so there is a draft transaction by default). 

For snapshots, I'm not entirely sure what to do with the “Share snapshot” link and then splitting up the “Save & Publish” buttons (which is a separate feature we need to look at too). These are the use cases I am thinking of:
- Share the exact current state of the Customizer settings, including dirty and non-dirty settings, a “full snapshot”.
- Share the modified (dirty) settings, for a “diff snapshot”.
- Allow unauthenticated user to access a snapshot to view on the frontend, or authenticated users to load the snapshot in the Customizer (maybe there should be an admin bar link for this instead of automatically redirecting).
- Do work in the Customizer and save your state as a draft for continuing to work on later (implies you want to update a previous snapshot/transaction).
- Remove “Publish” capability for unprivileged users, allowing others to just create snapshot/drafts of settings (diff or full). Privileged users can then access the snapshot (by URL, or by selecting from among a list of snapshot drafts?) to the publish.

So basically there are a lot of use cases here, and there is a lot of overlap among them, and so they are easy to get confused.

**Comment by valendesigns** [comment]
@westonruter That is certainly a lot of use cases that I hadn't considered. Good thing we're discussing it now while the code is still flexible.

> Share the exact current state of the Customizer settings, including dirty and non-dirty settings, a “full snapshot”.

This is currently how the PR is implemented.

> Share the modified (dirty) settings, for a “diff snapshot”.

This is possible, the only issue I see is that the snapshot would need to also save a flag to indicate whether the snapshot is a full or diff when previewing. And that flag would be used when creating or updating the snapshot.

> Allow unauthenticated user to access a snapshot to view on the frontend, or authenticated users to load the snapshot in the Customizer (maybe there should be an admin bar link for this instead of automatically redirecting).

So if an authenticated user goes to `http://sample.com/?customize_snapshot_uuid=65a9ab6a-9b8b-41b4-8ce1-6f511d61fbcb`, then we should amend the Customize link in the admin bar so that it removes `customize_snapshot_uuid` from the return url and adds it to the Customizer query vars. For example, the url above would generate `http://sample.com/wp-admin/customize.php?customize_snapshot_uuid=65a9ab6a-9b8b-41b4-8ce1-6f511d61fbcb&url=http%3A%2F%2Fsample.com%2F`.

> Do work in the Customizer and save your state as a draft for continuing to work on later (implies you want to update a previous snapshot/transaction).

This would mean that the share button needs to prompt the user and ask if you want to share the url or save a draft. If you save a draft then the URL dialog would display a Customizer url and not a front-end share url so you could load the snapshot in the customizer later and work on it when you're ready. Essentially the two are the same and the only difference is what the dialog displays because the draft would still be viewable by unauthenticated users.

> Remove “Publish” capability for unprivileged users, allowing others to just create snapshot/drafts of settings (diff or full). Privileged users can then access the snapshot (by URL, or by selecting from among a list of snapshot drafts?) to the publish.

I'm not sure how this would be implemented if the user doesn't have the capability to publish, how are they even viewing the Customizer?

**Comment by westonruter** [comment]
> This is possible, the only issue I see is that the snapshot would need to also save a flag to indicate whether the snapshot is a full or diff when previewing. And that flag would be used when creating or updating the snapshot.

Cool. We talked about this over Slack, and about storing all settings and the dirty state for each, so the mode (full or diff) can be decided later.

> So if an authenticated user goes to `http://sample.com/?customize_snapshot_uuid=65a9ab6a-9b8b-41b4-8ce1-6f511d61fbcb`, then we should amend the Customize link in the admin bar so that it removes `customize_snapshot_uuid` from the return url and adds it to the Customizer query vars. For example, the url above would generate `http://sample.com/wp-admin/customize.php?customize_snapshot_uuid=65a9ab6a-9b8b-41b4-8ce1-6f511d61fbcb&url=http%3A%2F%2Fsample.com%2F`.

Exactly! :+1:

> > Do work in the Customizer and save your state as a draft for continuing to work on later (implies you want to update a previous snapshot/transaction).
> 
> This would mean that the share button needs to prompt the user and ask if you want to share the url or save a draft. If you save a draft then the URL dialog would display a Customizer url and not a front-end share url so you could load the snapshot in the customizer later and work on it when you're ready. Essentially the two are the same and the only difference is what the dialog displays because the draft would still be viewable by unauthenticated users.

For the draft use case, let's omit that from consideration here. I think the draft functionality would be tied into a “Save Draft” button, which would be split out from “Save & Publish”, where “Publish” would then also be a separate button. Upon doing “Save Draft” it could then show the URL for the draft snapshot, and each time you hit that button it could update the same snapshot post: so the UUID sent along with the Customizer in the first load would be for a draft snapshot, and a UUID generated with each snapshot link share could be unique. A possibility.

> > Remove “Publish” capability for unprivileged users, allowing others to just create snapshot/drafts of settings (diff or full). Privileged users can then access the snapshot (by URL, or by selecting from among a list of snapshot drafts?) to the publish.
> 
> I'm not sure how this would be implemented if the user doesn't have the capability to publish, how are they even viewing the Customizer?

As discussed over Slack, this is out of scope for snapshots-as-sharable-links. But we can implement a discrete `customize_save` capability for saving the Customizer, and if the user doesn't have the cap we can hide the “Save & Publish” button and add this bit of code which blocks the request as well:

``` php
add_action( 'customize_save', function() {
    if ( ! current_user_can( 'customize_save' ) ) {
        status_header( 403 );
        wp_send_json_error( 'current_user_cannot' );
    }
} );
```

**Comment by valendesigns** [comment]
I guess `$apply_dirty` does need to be public because there is no way to set that from `update_snapshot()`. Also, would it make sense to mark all settings as dirty when doing a full snapshot?

**Comment by westonruter** [comment]
I think the in the case of updating (or inserting) a snapshot, it would not matter whether the snapshot was constructed with `$apply_dirty` as true or false. In both cases, the Customizer would be sending the full settings array with each setting's value and dirty state. So when doing a full snapshot vs a diff snapshot, the writing would be identical. It's only when reading that the `$apply_dirty` flag would apply, as that is when the `dirty` values are filtered out.

File: js/customize-snapshot.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+					html += '<p>' + _customizeWidgetsPlusCustomizeSnapshot.i18n.updateMsg + '</p>' +
+					'<input type="hidden" value="' + _customizeWidgetsPlusCustomizeSnapshot.scope + '" name="scope"></label>';
+				} else {
+					html += '<label for="type-0"><input id="type-0" type="radio" checked="checked" value="dirty" name="scope">' + _customizeWidgetsPlusCustomizeSnapshot.i18n.dirtyLabel + '</label><br>' +
+					'<label for="type-1"><input id="type-1" type="radio" value="full" name="scope">' + _customizeWidgetsPlusCustomizeSnapshot.i18n.fullLabel + '</label><br>';
+				}
+				html += '<input type="submit" tabindex="-1" style="position:absolute; top:-5000px">' +
+				'</fieldset>' +
+			'</form>' +
+		'</div>';
```
Comment (westonruter) [inline js/customize-snapshot.js]: We should use Underscore templates here. String concatenation of HTML will be an XSS red flag from VIP.

Comment (westonruter) [inline js/customize-snapshot.js]: See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/

File: php/class-customize-snapshot-manager.php [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+	/**
+	 * Save a snapshot.
+	 *
+	 * @param WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 * @param string $status The post status.
+	 * @return null|WP_Error
+	 */
+	public function save( \WP_Customize_Manager $manager, $status = 'draft' ) {
+		$new_setting_ids = array_diff( array_keys( $this->post_data ), array_keys( $manager->settings() ) );
+		$manager->add_dynamic_settings( wp_array_slice_assoc( $this->post_data, $new_setting_ids ) );
```
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: @westonruter Calling `add_dynamic_settings` here throws errors if you've added widgets or menus to the snapshot. I suppose it will throw errors for any dynamic settings. I'm not sure what's happening because the setting IDs are being passed correctly. I hacked Core to see what was being passed and the data looked fine while I was debugging but when you let it continue it throws an illegal index or null error from the setting ID.

Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: What errors?

Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: `Warning: Illegal offset type in isset or empty in /srv/www/wordpress-develop /src/wp-includes/class-wp-customize-manager.php on line 1063`

`Warning: Illegal offset type in isset or empty in /srv/www/wordpress-develop /src/wp-includes/class-wp-customize-widgets.php on line 118`

`Warning: preg_match() expects parameter 2 to be string, array given in /srv/www/wordpress-develop/src/wp-includes/class-wp-customize-widgets.php on line 122`

`Warning: preg_match() expects parameter 2 to be string, array given in /srv/www/wordpress-develop/src/wp-includes/class-wp-customize-nav-menus.php on line 403`

`Warning: preg_match() expects parameter 2 to be string, array given in /srv/www/wordpress-develop/src/wp-includes/class-wp-customize-nav-menus.php on line 407`

Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: These are the errors I get when I add a single new widget and try to update/save the snapshot.

**Comment by valendesigns** [comment]
@westonruter It appears that this feature is mostly working as we want it to for all users. Except, and this is an big one, with contextual settings (insert sad face here). We should connect tomorrow and discuss how to get the two working together.

At first glance I noticed that the Customizer thinks the settings are global, so when you save the snapshot data it doesn't differentiate the global and local scope. So even though previewing a snapshot with contextual settings will work, when you load it in the Customizer it sets the global scope for all the widgets to the values in the snapshot and `Localise to the current page` is unchecked. I know we keep slightly different hours, so feel free to checkout the PR tomorrow if you have a chance and haven't heard from me yet.

File: php/class-customize-snapshot-manager.php [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+			exit;
+		}
+	}
+
+	/**
+	 * Decode and store any $_POST['snapshot_customized'] data.
+	 *
+	 * The value is used by Customize_Snapshot_Manager::save().
+	 */
+	public function store_post_data() {
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php:142]: @valendesigns let's make sure that this method short-circuits never does anything if `! current_user_can( 'customize' )`

Comment (valendesigns) [inline php/class-customize-snapshot-manager.php:142]: Good call!

File: php/class-customize-snapshot-manager.php [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+	 * @param \WP_Customize_Manager $manager WP_Customize_Manager instance.
+	 */
+	public function save_snapshot( \WP_Customize_Manager $manager ) {
+		if ( ! current_user_can( 'customize' ) ) {
+			status_header( 403 );
+			wp_send_json_error( 'customize_not_allowed' );
+		} else if ( 'POST' !== $_SERVER['REQUEST_METHOD'] ) {
+			status_header( 405 );
+			wp_send_json_error( 'bad_method' );
+		}
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: @valendesigns since this `save_snapshot` method is tied to the `customize_save_after`, we already know that the user can `customize`, because if they are not authenticated they won't be able to get through `\WP_Customize_Manager::save()` to reach `customize_save_after` in any case.

File: php/class-customize-snapshot-manager.php [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+
+		$uuid = null;
+		if ( ! empty( $_POST['snapshot_uuid'] ) ) {
+			if ( empty( $this->post_data ) ) {
+				status_header( 400 );
+				wp_send_json_error( 'missing_snapshot_customized' );
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: @valendesigns instead of short-circuiting the `customize-save` request with an error, how about amending the `customize_save_response` filter to include these errors. In this way, any errors happening for snapshots won't cause other plugins from working properly. The `customize_save_response` filter is how general messages can be sent back.

File: php/class-customize-snapshot-manager.php [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+		} else if ( empty( $_POST['scope'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'invalid_customize_snapshot_scope' );
+		} else if ( empty( $this->post_data ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'missing_snapshot_customized' );
+		} else if ( empty( $_POST['preview'] ) ) {
+			status_header( 400 );
+			wp_send_json_error( 'missing_preview' );
+		}
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php:323]: An `if`/`else` of :muscle: _power_.

File: php/class-customize-snapshot.php [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+	/**
+	 * Persist the data in the snapshot post content.
+	 *
+	 * @param string $status
+	 *
+	 * @return null|\WP_Error
+	 */
+	public function save( $status = 'draft' ) {
+		if ( ! current_user_can( 'customize' ) ) {
+			return null;
```
Comment (westonruter) [inline php/class-customize-snapshot.php]: Should this return a `WP_Error` indicating the user is not authorized?

File: php/class-customize-snapshot-manager.php (hunk 5 above)
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: That is there in case it's called directly, but it could very well be removed.

File: php/class-customize-snapshot-manager.php [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+			if ( $this->can_preview( $setting, $this->post_data ) ) {
+				$post_data = $this->post_data[ $setting->id ];
+				$this->snapshot->set( $setting, $post_data['value'], $post_data['dirty'] );
+			}
+		}
+
+		$r = $this->snapshot->save( $status );
+		if ( is_wp_error( $r ) ) {
+			status_header( 500 );
+			wp_send_json_error( $r->get_error_message() );
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: @valendesigns instead of making `\CustomizeWidgetsPlus\Customize_Snapshot_Manager::save()` assume that it is being called in an Ajax request context, let's just have it return this `$r` right here. Then `\CustomizeWidgetsPlus\Customize_Snapshot_Manager::save_snapshot()` can handle the `WP_Error` and amend the `customize_save_response` with the error message for the Customizer JS to receive via `wp.customize.bind( 'saved', function ( response ) { ... } )`

File: php/class-customize-snapshot-manager.php (hunk 6 above)
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: I originally was going to use that filter, but I guess I got lazy in implementation and just mimicked `update_snapshot`.

File: js/customize-snapshot.js [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+			dialog.dialog( 'open' );
+			dialog.find( 'form input[name=scope]' ).blur();
+			dialog.next( '.ui-dialog-buttonpane' ).find( 'button' ).blur();
+		} );
+	};
+
+	/**
+	 * Make the AJAX request.
+	 */
+	self.doAjax = function() {
```
Comment (westonruter) [inline js/customize-snapshot.js]: Should this be more descriptive, like `sendUpdateSnapshotRequest()`?

File: php/class-customize-snapshot-manager.php (hunk 9 above)
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: I like!

File: php/class-customize-snapshot.php (hunk 8 above)
Comment (valendesigns) [inline php/class-customize-snapshot.php]: That would make more sense, yes.

File: php/class-customize-snapshot-manager.php (hunk 7 above)
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php:323]: Most of that came from your transaction patch... So that's you flexing those `if/else` :muscle:'s.

File: php/class-customize-snapshot-manager.php (hunk 4 above)
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php:142]: Actually, I take that back. This method redirects _all_ users who are trying to view a snapshot for an inactive theme. So a check for `current_user_can( 'customize' )` would fail to redirect. The redirect only removes the `customize_snapshot_uuid` & `scope` query args from the URL and redirects back on itself. So I don't see how this would be a concern?

File: js/customize-snapshot.js (hunk 10 above)
Comment (valendesigns) [inline js/customize-snapshot.js]: It wouldn't hurt anything. :smile:

File: php/class-customize-snapshot-manager.php (hunk 4 above)
Comment (westonruter) [inline php/class-customize-snapshot-manager.php:142]: I don't see where this method does a redirect. If the user is not authenticated, then we should not be reading anything from `$_POST`. If the user is not authenticated, we should only ever accept data read from the snapshot post, as the UUID keys into values that have been previously stored by an authenticated user.

Comment (valendesigns) [inline php/class-customize-snapshot-manager.php:142]: Sorry I thought you were talking about `maybe_force_redirect` but I see now we are discussing `store_post_data`... I agree, that should have an auth check. :+1:

File: php/class-customize-snapshot-manager.php [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+			'is_preview' => $this->snapshot->is_preview(),
+			'snapshot_theme' => $snapshot_theme,
+			'scope' => ( isset( $_GET['scope'] ) ? $_GET['scope'] : 'dirty' ),
+			'i18n' => array(
+				'saveButton' => __( 'Save', 'customize-widgets-plus' ),
+				'cancelButton' => __( 'Cancel', 'customize-widgets-plus' ),
+				'shareButton' => __( 'Share URL to preview', 'customize-widgets-plus' ),
+				'updateMsg' => __( 'Clicking "Save" will update the current snapshot.', 'customize-widgets-plus' ),
+				'errorMsg' => __( 'The snapshot could not be saved.', 'customize-widgets-plus' ),
+				'previewTitle' => __( 'Preview URL', 'customize-widgets-plus' ),
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: As noted in Slack, this isn't getting displayed after submitting the desired scope.

File: php/class-customize-snapshot-manager.php [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+
+	/**
+	 * Check if the setting can be previewed.
+	 *
+	 * @param \WP_Customize_Setting $setting A WP_Customize_Setting derived object.
+	 * @param array                 $values  All settings' values in the snapshot.
+	 * @return bool
+	 */
+	public function can_preview( $setting, $values ) {
+		if ( ! ( $setting instanceof \WP_Customize_Setting ) && ! is_subclass_of( $setting, 'WP_Customize_Setting' ) ) {
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: @valendesigns using `is_subclass_of` is redundant here, as `instanceof` also works for parent classes:

> instanceof can also be used to determine whether a variable is an instantiated object of a class that inherits from a parent class:
> http://php.net/manual/en/language.operators.type.php

File: php/class-customize-snapshot-manager.php [hunk 13]
```diff
@@ -0,0 +1,559 @@
...
+	 * @param \WP_Admin_Bar $wp_admin_bar WP_Admin_Bar instance.
+	 */
+	public function customize_menu( $wp_admin_bar ) {
+		// Don't show for users who can't access the customizer or when in the admin.
+		if ( ! current_user_can( 'customize' ) || is_admin() ) {
+			return;
+		}
+
+		$current_url = $this->current_url();
+		parse_str( parse_url( $current_url, PHP_URL_QUERY ), $query_vars );
```
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: @valendesigns this looks unsafe. Instead of parsing query vars from the current URL, why not just use `$_GET`?

Also, more importantly, these aren't being sanitized.

Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: Ah, well UUID is sanitized, but `scope` is not.

File: php/class-customize-snapshot-manager.php (hunk 12 above)
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: Well... :disappointed:

File: php/class-customize-snapshot-manager.php (hunk 13 above)
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: Sounds sane, I'll fix it!

**Comment by valendesigns** [comment]
Development has officially moved to https://github.com/xwp/wp-customize-snapshots
